File: app/src/main/java/com/hlibrary/base/fragment/BaseFragment.java
package com.hlibrary.base.fragment;

import android.databinding.DataBindingUtil;
import android.databinding.ViewDataBinding;
import android.os.Bundle;
import android.support.annotation.LayoutRes;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentActivity;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.support.v7.widget.Toolbar;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.hlibrary.R;
import com.hlibrary.base.CommonHandler;
import com.hlibrary.base.ITabUpdate;
import com.hlibrary.base.activity.BaseActivity;

public class BaseFragment<T extends ViewDataBinding> extends Fragment {

    private static final int CONTENT_ID = R.id.frmlyt_tab;

    protected CommonHandler mFragmentHandler;
    protected T mViewDataBind;
    protected Bundle savedState;

    protected String title = "BaseFragment";

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setRetainInstance(true);
    }

    @Override
    public final void onActivityCreated(Bundle savedInstanceState) {
        onLocalActivityCreated(savedInstanceState);
        super.onActivityCreated(savedInstanceState);
        mFragmentHandler = CommonHandler.Companion.obtain(getActivity());
        if (!restoreStateFromArguments()) {
            // First Time, Initialize something here
            onFirstTimeLaunched();
        }

    }

    @Override
    public void onDestroy() {
        mFragmentHandler.release();
        super.onDestroy();
    }

    public void onLocalActivityCreated(Bundle savedInstanceState) {
    }

    protected View findViewById(int id) {
        return getView().findViewById(id);
    }

    protected void onFirstTimeLaunched() {

    }

    protected final View setContentView(@LayoutRes int layoutResID, @NonNull LayoutInflater inflater, @NonNull
            ViewGroup container) {
        mViewDataBind = DataBindingUtil.inflate(inflater, layoutResID, container, false);
        return mViewDataBind.getRoot();
    }

    protected void onBtnLeftClick(View v) {
    }

    protected void onBtnRightClick(View v) {
    }

    protected void setupToolbar(Toolbar toolbar) {
        BaseActivity baseActivity = (BaseActivity) getActivity();
        baseActivity.setupToolbar(toolbar);
    }

    @Override
    public final void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        saveStateToArguments(outState);
    }

    private static final String STATE_TAG = "STATE_TAG";

    private void saveStateToArguments(Bundle outState) {
        savedState = saveState();
        if (savedState != null) {
            Bundle b = getArguments();
            if (b == null)
                b = new Bundle();
            b.putBundle(STATE_TAG, savedState);
        }
    }

    private boolean restoreStateFromArguments() {
        Bundle b = getArguments();
        if (b != null) {
            if (b.containsKey(STATE_TAG)) {
                savedState = b.getBundle(STATE_TAG);
                if (savedState != null) {
                    restoreState();
                    return true;
                }
            }
        }
        return false;
    }

    private void restoreState() {
        if (savedState != null) {
            // For Example
            //tv1.setText(savedState.getString(text));
            onRestoreState(savedState);
        }
    }

    protected void onRestoreState(Bundle savedInstanceState) {

    }

    private Bundle saveState() {
        Bundle state = new Bundle();
        // For Example
        //state.putString(text, tv1.getText().toString());
        onSaveState(state);
        return state;
    }

    protected void onSaveState(Bundle outState) {

    }

    public static <T extends BaseFragment> T getFragment(
            FragmentManager fragmentManager, Class<T> cls) {
        @SuppressWarnings("unchecked")
        T fragment = (T) fragmentManager.findFragmentByTag(cls.getName());
        return fragment;
    }

    public static void isAdd(int id, FragmentTransaction fragmentTransaction,
                             Fragment fragment) {
        if (!fragment.isAdded()) {
            fragmentTransaction
                    .add(id, fragment, fragment.getClass().getName());
            fragmentTransaction.hide(fragment);
        }
    }

    private static String curKey;

    public static void showAndHideFragment(int id,
                                           FragmentManager fragmentManager, Fragment fragment) {
        FragmentTransaction fragmentTransaction = fragmentManager
                .beginTransaction();

        if (curKey != null) {
            Fragment curFragment = fragmentManager.findFragmentByTag(curKey);
            if (curFragment != null)
                fragmentTransaction.hide(curFragment);
        }
        curKey = fragment.getClass().getName();
        if (fragment.isAdded()) {
            fragmentTransaction.show(fragment);
        } else {
            fragmentTransaction.add(id, fragment, curKey);
        }
        if (!fragmentTransaction.isEmpty()) {
            fragmentTransaction.commit();
            fragmentManager.executePendingTransactions();
        }
    }

    public static void showAndHideFragment(FragmentManager fragmentManager, Fragment fragment) {
        showAndHideFragment(CONTENT_ID, fragmentManager, fragment);
    }

    public static void showFragment(int id, FragmentManager fragmentManager,
                                    Fragment fragment) {
        FragmentTransaction fragmentTransaction = fragmentManager
                .beginTransaction();
        fragmentTransaction.replace(id, fragment);
        if (!fragmentTransaction.isEmpty()) {
            fragmentTransaction.commit();
            fragmentManager.executePendingTransactions();
        }
    }

    public static void showFragment(FragmentManager fragmentManager, Fragment fragment) {
        showFragment(CONTENT_ID, fragmentManager, fragment);
    }

    public void onClick(View v) {
    }

    public int getColor(int resId) {
        return getResources().getColor(resId);
    }

    public int getDimension(int id) {
        return getResources().getDimensionPixelSize(id);
    }

    public String getTitle() {
        return title;
    }

    private ITabUpdate tabUpdate;

    public void onUpdateTab(int index) {
        if (tabUpdate != null) {
            tabUpdate.onUpdateTab(index);
        } else {
            FragmentActivity activity = getActivity();
            if (activity instanceof ITabUpdate) {
                tabUpdate = (ITabUpdate) activity;
                tabUpdate.onUpdateTab(index);
            }
        }
    }

}

File: app/src/main/java/com/hlibrary/base/fragment/BaseFragment.kt
package com.hlibrary.base.fragment

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.LayoutRes
import androidx.appcompat.widget.Toolbar
import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentTransaction
import com.hlibrary.R
import com.hlibrary.base.ITabUpdate
import com.hlibrary.base.activity.BaseActivity

open class BaseFragment<T : ViewDataBinding?> : Fragment() {

    protected var mViewDataBind: T? = null
    protected var savedState: Bundle? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        retainInstance = true
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        onLocalActivityCreated(savedInstanceState)
        super.onActivityCreated(savedInstanceState)
        if (!restoreStateFromArguments()) { // First Time, Initialize something here
            onFirstTimeLaunched()
        }
    }

    fun onLocalActivityCreated(savedInstanceState: Bundle?) {}


    protected fun onFirstTimeLaunched() {}

    protected fun setContentView(@LayoutRes layoutResID: Int, inflater: LayoutInflater, container: ViewGroup): View? {
        mViewDataBind = DataBindingUtil.inflate<T>(inflater, layoutResID, container, false)
        return mViewDataBind?.root
    }

    protected fun setupToolbar(toolbar: Toolbar) {
        val baseActivity = activity as BaseActivity<*>?
        baseActivity?.setupToolbar(toolbar)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        saveStateToArguments(outState)
    }

    private fun saveStateToArguments(outState: Bundle) {
        savedState = saveState()
        if (savedState != null) {
            var b = arguments
            if (b == null) {
                b = Bundle()
            }
            b.putBundle(STATE_TAG, savedState)
        }
    }

    private fun restoreStateFromArguments(): Boolean {
        val b = arguments
        if (b != null) {
            if (b.containsKey(STATE_TAG)) {
                savedState = b.getBundle(STATE_TAG)
                if (savedState != null) {
                    restoreState()
                    return true
                }
            }
        }
        return false
    }

    private fun restoreState() {
        if (savedState != null) {
            onRestoreState(savedState)
        }
    }

    protected fun onRestoreState(savedInstanceState: Bundle?) {}

    private fun saveState(): Bundle {
        val state = Bundle()
        onSaveState(state)
        return state
    }

    protected fun onSaveState(outState: Bundle?) {}



    private var tabUpdate: ITabUpdate? = null

    fun onUpdateTab(index: Int) {
        if (tabUpdate != null) {
            tabUpdate?.onUpdateTab(index)
        } else {
            val activity = activity
            if (activity is ITabUpdate) {
                tabUpdate = activity
                tabUpdate?.onUpdateTab(index)
            }
        }
    }

    companion object {
        private val CONTENT_ID = R.id.frmlyt_tab
        private const val STATE_TAG = "STATE_TAG"
        fun <T : BaseFragment<*>?> getFragment(
                fragmentManager: FragmentManager, cls: Class<T>): T? {
            return fragmentManager.findFragmentByTag(cls.name) as T?
        }

        fun isAdd(id: Int, fragmentTransaction: FragmentTransaction,
                  fragment: Fragment) {
            if (!fragment.isAdded) {
                fragmentTransaction
                        .add(id, fragment, fragment.javaClass.name)
                fragmentTransaction.hide(fragment)
            }
        }

        private var curKey: String? = null

        fun showAndHideFragment(id: Int,
                                fragmentManager: FragmentManager, fragment: Fragment) {
            val fragmentTransaction = fragmentManager
                    .beginTransaction()
            if (curKey != null) {
                val curFragment = fragmentManager.findFragmentByTag(curKey)
                if (curFragment != null) {
                    fragmentTransaction.hide(curFragment)
                }
            }
            curKey = fragment.javaClass.name
            if (fragment.isAdded) {
                fragmentTransaction.show(fragment)
            } else {
                fragmentTransaction.add(id, fragment, curKey)
            }
            if (!fragmentTransaction.isEmpty) {
                fragmentTransaction.commit()
                fragmentManager.executePendingTransactions()
            }
        }

        fun showAndHideFragment(fragmentManager: FragmentManager, fragment: Fragment) {
            showAndHideFragment(CONTENT_ID, fragmentManager, fragment)
        }

        fun showFragment(id: Int, fragmentManager: FragmentManager,
                         fragment: Fragment?) {
            val fragmentTransaction = fragmentManager
                    .beginTransaction()
            fragmentTransaction.replace(id, fragment!!)
            if (!fragmentTransaction.isEmpty) {
                fragmentTransaction.commit()
                fragmentManager.executePendingTransactions()
            }
        }

        fun showFragment(fragmentManager: FragmentManager, fragment: Fragment?) {
            showFragment(CONTENT_ID, fragmentManager, fragment)
        }
    }
}